x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,369 @@
+                                        + TextExpansionResults.NAME
+                                        + "] received ["
+                                        + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                        + "]. Is ["
+                                        + modelId
+                                        + "] a compatible model?"
+                                )
+                            );
+                        }
+                    }, listener::onFailure)
+                )
+            );
+
+            return new SparseVectorQueryBuilder(this, textExpansionResultsSupplier);
+
+        } else {
+            return vectorDimensionsToQuery(fieldName, vectorDimensions);
+        }
+
+    }
+
+    private QueryBuilder vectorDimensionsToQuery(String fieldName, List<VectorDimension> vectorDimensions) {
+        if (tokenPruningConfig != null) {
+            VectorDimensionsQueryBuilder vectorDimensionsQueryBuilder = new VectorDimensionsQueryBuilder(
+                fieldName,
+                vectorDimensions,
+                tokenPruningConfig
+            );
+            vectorDimensionsQueryBuilder.queryName(queryName);
+            vectorDimensionsQueryBuilder.boost(boost);
+            return vectorDimensionsQueryBuilder;
+        }
+        // Note: Weighted tokens queries were introduced in 8.13.0. To support mixed version clusters prior to 8.13.0,
+        // if no token pruning configuration is specified we fall back to a boolean query.
+        // TODO this should be updated to always use a WeightedTokensQueryBuilder once it's in all supported versions.
+        var boolQuery = QueryBuilders.boolQuery();
+        for (var weightedToken : vectorDimensions) {
+            boolQuery.should(QueryBuilders.termQuery(fieldName, weightedToken.token()).boost(weightedToken.weight()));
+        }
+        boolQuery.minimumShouldMatch(1);
+        boolQuery.boost(boost);
+        boolQuery.queryName(queryName);
+        return boolQuery;
+    }
+
+    private QueryBuilder textExpansionResultsToQuery(String fieldName, TextExpansionResults textExpansionResults) {
+        return vectorDimensionsToQuery(fieldName, textExpansionResults.getVectorDimensions());
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) {
+        throw new IllegalStateException("sparse_vector should have been rewritten to another query type");
+    }
+
+    @Override
+    protected boolean doEquals(SparseVectorQueryBuilder other) {
+        return Objects.equals(fieldName, other.fieldName)
+            && Objects.equals(modelText, other.modelText)
+            && Objects.equals(modelId, other.modelId)
+            && Objects.equals(tokenPruningConfig, other.tokenPruningConfig)
+            && Objects.equals(vectorDimensions, other.vectorDimensions)
+            && Objects.equals(vectorDimensionsSupplier, other.vectorDimensionsSupplier);
```

Comment:
`SetOnce` doesn't implement `equals`, so this essentially checks that the query builders are the same instance.